syntax = "proto3";

package arduino_trader.portfolio;

import "common/common.proto";
import "common/position.proto";
import "common/security.proto";

// Portfolio Service - Handles portfolio state and positions
service PortfolioService {
  // Get current portfolio positions
  rpc GetPositions(GetPositionsRequest) returns (GetPositionsResponse);

  // Get a specific position
  rpc GetPosition(GetPositionRequest) returns (GetPositionResponse);

  // Get portfolio summary
  rpc GetSummary(GetSummaryRequest) returns (GetSummaryResponse);

  // Get portfolio performance
  rpc GetPerformance(GetPerformanceRequest) returns (GetPerformanceResponse);

  // Update positions (sync from broker)
  rpc UpdatePositions(UpdatePositionsRequest) returns (UpdatePositionsResponse);

  // Get cash balance
  rpc GetCashBalance(GetCashBalanceRequest) returns (GetCashBalanceResponse);

  // Health check
  rpc HealthCheck(common.Empty) returns (common.HealthCheckResponse);
}

// Get positions request
message GetPositionsRequest {
  string account_id = 1;
  bool include_closed = 2;
}

// Get positions response
message GetPositionsResponse {
  repeated common.Position positions = 1;
  int32 total_positions = 2;
}

// Get position request
message GetPositionRequest {
  string account_id = 1;
  string isin = 2;
}

// Get position response
message GetPositionResponse {
  bool found = 1;
  common.Position position = 2;
}

// Get summary request
message GetSummaryRequest {
  string account_id = 1;
}

// Get summary response
message GetSummaryResponse {
  string portfolio_hash = 1;
  common.Money total_value = 2;
  common.Money total_cost = 3;
  common.Money total_pnl = 4;
  double total_pnl_pct = 5;
  common.Money cash_balance = 6;
  int32 position_count = 7;
  common.Timestamp last_updated = 8;
  map<string, double> sector_allocation = 9;
  map<string, double> region_allocation = 10;
}

// Get performance request
message GetPerformanceRequest {
  string account_id = 1;
  int32 days = 2;
}

// Get performance response
message GetPerformanceResponse {
  repeated PerformanceDataPoint history = 1;
  PerformanceMetrics metrics = 2;
}

// Performance data point
message PerformanceDataPoint {
  common.Timestamp date = 1;
  common.Money value = 2;
  double return_pct = 3;
}

// Performance metrics
message PerformanceMetrics {
  double total_return_pct = 1;
  double annualized_return_pct = 2;
  double volatility = 3;
  double sharpe_ratio = 4;
  double max_drawdown_pct = 5;
  common.Money peak_value = 6;
  common.Timestamp peak_date = 7;
}

// Update positions request
message UpdatePositionsRequest {
  string account_id = 1;
  repeated common.Position positions = 2;
  bool force_sync = 3;
}

// Update positions response
message UpdatePositionsResponse {
  bool success = 1;
  int32 positions_updated = 2;
  int32 positions_added = 3;
  int32 positions_removed = 4;
  common.Timestamp sync_time = 5;
}

// Get cash balance request
message GetCashBalanceRequest {
  string account_id = 1;
}

// Get cash balance response
message GetCashBalanceResponse {
  common.Money cash_balance = 1;
  common.Money pending_deposits = 2;
  common.Money pending_withdrawals = 3;
  common.Money available_for_trading = 4;
}
